src/main/java/org/zalando/nakadi/service/timeline/TimelineSyncImpl.java

```diff
@@ -0,0 +1,333 @@
+    private final BlockingQueue<DelayedChange> queuedChanges = new LinkedBlockingQueue<>();
+    private final AtomicBoolean newVersionPresent = new AtomicBoolean(true);
+
+    @Autowired
+    public TimelineSyncImpl(final ZooKeeperHolder zooKeeperHolder, final UUIDGenerator uuidGenerator)
+            throws InterruptedException {
+        this.thisId = uuidGenerator.randomUUID().toString();
+        this.zooKeeperHolder = zooKeeperHolder;
+        this.initializeZkStructure();
+    }
+
+    @Override
+    public String getNodeId() {
+        return thisId;
+    }
+
+    private String toZkPath(final String path) {
+        return ROOT_PATH + path;
+    }
+
+    private void initializeZkStructure() throws InterruptedException {
+        LOG.info("Starting initialization");
+        runLocked(() -> {
+            try {
+                try {
+                    // 1. Create version node, if needed, keep in mind that lock built root path
+                    zooKeeperHolder.get().create().withMode(CreateMode.PERSISTENT)
+                            .forPath(toZkPath("/version"), "0".getBytes(Charsets.UTF_8));
+
+                    // 2. Create locked event types structure
+                    zooKeeperHolder.get().create().withMode(CreateMode.PERSISTENT)
+                            .forPath(toZkPath("/locked_et"), "[]".getBytes(Charsets.UTF_8));
+                    // 3. Create nodes root path
+                    zooKeeperHolder.get().create().withMode(CreateMode.PERSISTENT)
+                            .forPath(toZkPath("/nodes"), "[]".getBytes(Charsets.UTF_8));
+                } catch (final KeeperException.NodeExistsException ignore) {
+                }
+
+                LOG.info("Registering node in zk");
+                zooKeeperHolder.get().create().withMode(CreateMode.EPHEMERAL)
+                        .forPath(toZkPath("/nodes/" + thisId), "0".getBytes(Charsets.UTF_8));
+
+            } catch (final Exception e) {
+                LOG.error("Failed to initialize timeline synchronization", e);
+                throw new RuntimeException(e);
+            }
+        });
+        // 4. React on what was received and write node version to zk.
+        reactOnEventTypesChange();
+    }
+
+    private <T> T readData(final String relativeName,
+                           @Nullable final Watcher watcher, final Function<String, T> converter) throws Exception {
+        final byte[] data;
+        final String zkPath = toZkPath(relativeName);
+        if (null == watcher) {
+            data = zooKeeperHolder.get().getData().forPath(zkPath);
+        } else {
+            data = zooKeeperHolder.get().getData().usingWatcher(watcher).forPath(zkPath);
+        }
+        return converter.apply(new String(data, Charsets.UTF_8));
+    }
+
+    @Scheduled(fixedDelay = 500)
+    public void reactOnEventTypesChange() throws InterruptedException {
+        checkForNewChange();
+        while (null != queuedChanges.peek()) {
```

Comment:
Forget about it... it's scheduled up there. It got lost in the middle of comments.